core/node/provider.go

```diff
@@ -123,49 +241,262 @@ We observed that you recently provided %d keys at an average rate of %v per key.
+			dhtprovider.WithReprovideInterval(reprovideInterval),
+			dhtprovider.WithMaxReprovideDelay(time.Hour),
+			dhtprovider.WithOfflineDelay(cfg.Reprovider.Sweep.OfflineDelay.WithDefault(config.DefaultReproviderSweepOfflineDelay)),
+			dhtprovider.WithConnectivityCheckOnlineInterval(1 * time.Minute),
+
+			dhtprovider.WithMaxWorkers(int(cfg.Reprovider.Sweep.MaxWorkers.WithDefault(config.DefaultReproviderSweepMaxWorkers))),
+			dhtprovider.WithDedicatedPeriodicWorkers(int(cfg.Reprovider.Sweep.DedicatedPeriodicWorkers.WithDefault(config.DefaultReproviderSweepDedicatedPeriodicWorkers))),
+			dhtprovider.WithDedicatedBurstWorkers(int(cfg.Reprovider.Sweep.DedicatedBurstWorkers.WithDefault(config.DefaultReproviderSweepDedicatedBurstWorkers))),
+			dhtprovider.WithMaxProvideConnsPerWorker(int(cfg.Reprovider.Sweep.MaxProvideConnsPerWorker.WithDefault(config.DefaultReproviderSweepMaxProvideConnsPerWorker))),
+		}
+
+		prov, err := dhtprovider.New(opts...)
+		return prov, keyStore, err
+	})
+
+	type keystoreInput struct {
+		fx.In
+		Provider    DHTProvider
+		KeyStore    *rds.KeyStore
+		KeyProvider provider.KeyChanFunc
+	}
+	initKeyStore := fx.Invoke(func(lc fx.Lifecycle, in keystoreInput) {
+		var (
+			cancel context.CancelFunc
+			done   = make(chan struct{})
+		)
+
+		syncKeyStore := func(ctx context.Context) error {
+			kcf, err := in.KeyProvider(ctx)
+			if err != nil {
+				return err
+			}
+			if err := in.KeyStore.ResetCids(ctx, kcf); err != nil {
+				return err
+			}
+			if err := in.Provider.RefreshSchedule(); err != nil {
+				logger.Infow("refreshing provider schedule", "err", err)
+			}
+			return nil
 		}
 
 		lc.Append(fx.Hook{
+			OnStart: func(ctx context.Context) error {
+				// Set the KeyProvider as a garbage collection function for the
+				// keystore. Periodically purge the KeyStore from all its keys and
+				// replace them with the keys that needs to be reprovided, coming from
+				// the KeyChanFunc. So far, this is the less worse way to remove CIDs
+				// that shouldn't be reprovided from the provider's state.
+				if err := syncKeyStore(ctx); err != nil {
+					return err
+				}
+
+				gcCtx, c := context.WithCancel(context.Background())
+				cancel = c
+
+				go func() { // garbage collection loop for cids to reprovide
+					defer close(done)
+					ticker := time.NewTicker(reprovideInterval)
+					defer ticker.Stop()
+
+					for {
+						select {
+						case <-gcCtx.Done():
+							return
+						case <-ticker.C:
+							if err := syncKeyStore(gcCtx); err != nil {
+								logger.Errorw("provider keystore sync", "err", err)
+							}
+						}
+					}
+				}()
+				return nil
+			},
 			OnStop: func(ctx context.Context) error {
-				return sys.Close()
+				if cancel != nil {
+					cancel()
+				}
+				select {
+				case <-done:
+				case <-ctx.Done():
+					return ctx.Err()
+				}
+				// KeyStore state isn't be persisted across restarts.
+				if err := in.KeyStore.Empty(ctx); err != nil {
+					return err
```

Comment:
Do we still want to try to close the KeyStore here?
 ```suggestion
 					in.KeyStore.Close()
					return err
```

Reply:
Simply logged the error since it isn't critical if to empty the KeyStore. 

Also, it won't be necessary to close the KeyStore after https://github.com/libp2p/go-libp2p-kad-dht/pull/1140